In [ ]:
# Cell 1: Install & import everything
!pip install tensorflow matplotlib seaborn --quiet

import tensorflow as tf
from tensorflow.keras import layers, models
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import classification_report, confusion_matrix
import time

print("TensorFlow version:", tf.__version__)

TensorFlow version: 2.19.0


In [ ]:
# Cell 2: Load Fashion-MNIST
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.fashion_mnist.load_data()

print("Dataset loaded successfully!")
print("Train shape:", x_train.shape, "Labels:", y_train.shape)
print("Test shape:", x_test.shape, "Labels:", y_test.shape)

29515/29515 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
26421880/26421880 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
5148/5148 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
4422102/4422102 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step
Dataset loaded successfully!
Train shape: (60000, 28, 28) Labels: (60000,)
Test shape: (10000, 28, 28) Labels: (10000,)


In [ ]:
# Cell 3: Preprocess
x_train = x_train / 255.0
x_test  = x_test  / 255.0

# Add channel dimension for CNN (28,28) → (28,28,1)
x_train = np.expand_dims(x_train, axis=-1)
x_test  = np.expand_dims(x_test,  axis=-1)

print("Preprocessing done. New shape:", x_train.shape)

Preprocessing done. New shape: (60000, 28, 28, 1)


In [ ]:
# Cell 4: Build CNN
model = models.Sequential([
    layers.Conv2D(32, (3,3), activation='relu', input_shape=(28,28,1)),
    layers.MaxPooling2D((2,2)),
    layers.Conv2D(64, (3,3), activation='relu'),
    layers.MaxPooling2D((2,2)),
    layers.Flatten(),
    layers.Dense(128, activation='relu'),
    layers.Dropout(0.5),
    layers.Dense(10, activation='softmax')
])

model.summary()

/usr/local/lib/python3.12/dist-packages/keras/src/layers/convolutional/base_conv.py:113: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 26, 26, 32)     │           320 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 13, 13, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 11, 11, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 5, 5, 64)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 1600)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │       204,928 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 225,034 (879.04 KB)

 Trainable params: 225,034 (879.04 KB)

 Non-trainable params: 0 (0.00 B)

In [ ]:
# Cell 5: Compile and train
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

start_time = time.time()
history = model.fit(x_train, y_train,
                    epochs=20,
                    batch_size=128,
                    validation_split=0.1,
                    verbose=1)
training_time = time.time() - start_time
print(f"Training completed in {training_time:.2f} seconds")

Epoch 1/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 45s 104ms/step - accuracy: 0.7538 - loss: 0.6888 - val_accuracy: 0.8447 - val_loss: 0.4139
Epoch 2/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 77s 92ms/step - accuracy: 0.8466 - loss: 0.4293 - val_accuracy: 0.8697 - val_loss: 0.3462
Epoch 3/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 37s 88ms/step - accuracy: 0.8672 - loss: 0.3733 - val_accuracy: 0.8828 - val_loss: 0.3078
Epoch 4/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 38s 90ms/step - accuracy: 0.8765 - loss: 0.3417 - val_accuracy: 0.8885 - val_loss: 0.2986
Epoch 5/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 39s 91ms/step - accuracy: 0.8855 - loss: 0.3191 - val_accuracy: 0.8957 - val_loss: 0.2771
Epoch 6/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 38s 90ms/step - accuracy: 0.8916 - loss: 0.2978 - val_accuracy: 0.8990 - val_loss: 0.2704
Epoch 7/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 38s 91ms/step - accuracy: 0.8970 - loss: 0.2860 - val_accuracy: 0.9052 - val_loss: 0.2578
Epoch 8/20
422/422 ━━━━━━━━━━━━━━━━━━━━ 38s 90ms/step - accuracy: 0.9035 - loss: 0.2692 -

In [ ]:
# Cell 6: Evaluation
test_loss, test_acc = model.evaluate(x_test, y_test, verbose=0)
print(f"Test Accuracy: {test_acc*100:.2f}%")

# Predictions
y_pred = np.argmax(model.predict(x_test), axis=1)

# Classification report
print(classification_report(y_test, y_pred, target_names=[
    "T-shirt/top","Trouser","Pullover","Dress","Coat",
    "Sandal","Shirt","Sneaker","Bag","Ankle boot"]))

Test Accuracy: 91.26%
313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 7ms/step
              precision    recall  f1-score   support

 T-shirt/top       0.86      0.87      0.87      1000
     Trouser       0.99      0.98      0.98      1000
    Pullover       0.88      0.86      0.87      1000
       Dress       0.93      0.89      0.91      1000
        Coat       0.83      0.90      0.86      1000
      Sandal       0.99      0.96      0.98      1000
       Shirt       0.76      0.74      0.75      1000
     Sneaker       0.94      0.98      0.96      1000
         Bag       0.98      0.98      0.98      1000
  Ankle boot       0.97      0.96      0.96      1000

    accuracy                           0.91     10000
   macro avg       0.91      0.91      0.91     10000
weighted avg       0.91      0.91      0.91     10000



## 6. Performance Analysis

### Interpretation of Results

From the evaluation metrics and visualizations, we can conclude the following:

*   **Overall Accuracy:** The model achieved a test accuracy of **91.26%**, which is quite good for the Fashion-MNIST dataset.
*   **Training History:**
    *   The **accuracy plots** show that both training and validation accuracy increased steadily, indicating that the model was learning effectively. The validation accuracy generally tracked closely with the training accuracy, suggesting good generalization.
    *   The **loss plots** show a similar trend, with both training and validation loss decreasing over epochs. Towards the end, there might be a slight divergence or flattening, which could indicate the model is starting to converge or perhaps lightly overfitting, though the validation loss remains low.
*   **Classification Report:**
    *   **High Precision/Recall/F1-score** are observed for classes like 'Trouser', 'Sandal', 'Sneaker', 'Bag', and 'Ankle boot'. This means the model is very good at identifying these items and rarely confuses them with others.
    *   **Lower Performance for 'Shirt' and 'Pullover'**: The 'Shirt' class has the lowest F1-score (0.75), indicating that it is often misclassified. 'Pullover' and 'Coat' also have slightly lower scores compared to other classes.
*   **Confusion Matrix Insights:**
    *   The confusion matrix visually confirms the strengths and weaknesses. It highlights that 'Shirt' is frequently confused with 'T-shirt/top', 'Pullover', and 'Coat', which are visually similar clothing items. Similarly, 'Pullover' and 'Coat' might be confused with each other.
    *   There are very few misclassifications between distinct categories like 'Sandal' and 'Trouser', as expected.

In summary, the CNN model performs well on the Fashion-MNIST dataset, demonstrating strong generalization capabilities. The main areas for potential improvement lie in distinguishing between visually similar clothing items, particularly 'Shirt' from 'T-shirt/top', 'Pullover', and 'Coat'.

## 7. Hyperparameter Tuning

Hyperparameter tuning is the process of finding the optimal set of hyperparameters for a machine learning model to achieve the best performance. Instead of manually guessing, we can use automated methods to search through a defined range of hyperparameter values.

For this CNN model, we can tune hyperparameters such as the number of filters in convolutional layers, kernel sizes, number of dense units, dropout rate, and learning rate. We will use `keras_tuner` for this task.

### Impact of Hyperparameter Tuning

After running the hyperparameter tuning process, we observe the following impacts:

*   **Improved Performance (Potentially):** The `best_model` derived from tuning typically achieves a validation accuracy that is equal to or better than the baseline model, as the tuner explores various combinations to find an optimal set. The reported `Best Model Test Accuracy` indicates the performance of the tuned model.
*   **Identification of Optimal Configuration:** The tuning process provides insights into which hyperparameter values work best for this specific dataset and model architecture. For example, it suggests optimal filter sizes, dense layer units, dropout rates, and learning rates.
*   **Reduced Overfitting/Underfitting:** By adjusting parameters like dropout rate and learning rate, tuning can help in achieving a better balance between bias and variance, potentially leading to a model that generalizes better to unseen data and avoids both severe overfitting and underfitting.
*   **Computational Cost:** Hyperparameter tuning is computationally intensive, as it involves training multiple models. The `max_trials` and `executions_per_trial` parameters directly influence the time and resources required.

In summary, hyperparameter tuning is a crucial step to maximize the model's performance by systematically exploring the hyperparameter space. The optimal hyperparameters found provide a more robust and better-performing model compared to one built with arbitrary parameter choices.